### PR TITLE
fix writing gadget file

### PR DESCRIPTION
**Pull request by uweschmitt**
The files written using the gadget file writer were not compatible with `Pylians` and also caused issues with `Gadget 2` itself.

This pull request fixes this.

### CODE REVIEW

**Comment by warmtoday** [comment]
您的邮件已收到，谢谢！

**Comment by apetri** [comment]
Are unit tests passing? Travis CI is not active anymore

File: lenstools/extern/write_gadget_particles.c [hunk 1]
```diff
@@ -1,96 +1,102 @@
 #include <stdio.h>
 #include <unistd.h>
+#include <unistd.h>
```
Comment (apetri) [inline lenstools/extern/write_gadget_particles.c:3]: why this double include?

Comment (uweschmitt) [reply lenstools/extern/write_gadget_particles.c:3]: I'm cleaning up. I found another issue with gadget2 rejecting files. Will ping you here when I'm done.

Comment (apetri) [reply lenstools/extern/write_gadget_particles.c:3]: got it; just run unit tests offline to make sure they pass

**Comment by uweschmitt** [comment]
So, I digged deeper and figured out that current `gadget2` and `lenstools` don't implement the same layout for the file header.

From gadget2:

```
extern struct io_header
{
  int npart[6];                        /*!< number of particles of each type in this file */
  double mass[6];                      /*!< mass of particles of each type. If 0, then the masses are explicitly
                                            stored in the mass-block of the snapshot file, otherwise they are omitted */
  double time;                         /*!< time of snapshot file */
  double redshift;                     /*!< redshift of snapshot file */
  int flag_sfr;                        /*!< flags whether the simulation was including star formation */
  int flag_feedback;                   /*!< flags whether feedback was included (obsolete) */
  unsigned int npartTotal[6];          /*!< total number of particles of each type in this snapshot. This can be
                                            different from npart if one is dealing with a multi-file snapshot. */
  int flag_cooling;                    /*!< flags whether cooling was included  */
  int num_files;                       /*!< number of files in multi-file snapshot */
  double BoxSize;                      /*!< box-size of simulation in case periodic boundaries were used */
  double Omega0;                       /*!< matter density in units of critical density */
  double OmegaLambda;                  /*!< cosmological constant parameter */
  double HubbleParam;                  /*!< Hubble parameter in units of 100 km/sec/Mpc */
  int flag_stellarage;                 /*!< flags whether the file contains formation times of star particles */
  int flag_metals;                     /*!< flags whether the file contains metallicity values for gas and star particles */
  unsigned int npartTotalHighWord[6];  /*!< High word of the total number of particles of each type */
  int  flag_entropy_instead_u;         /*!< flags that IC-file contains entropy instead of u */
  char fill[60];>---               /*!< fills to 256 Bytes */
}
```

and `lenstools` uses

```
struct io_header_1
{
  int npart[6];
  double mass[6];
  double time;
  double redshift;
  int flag_sfr;
  int flag_feedback;
  int npartTotal[6];
  int flag_cooling;
  int num_files;
  double BoxSize;
  double Omega0;
  double OmegaLambda;
  double HubbleParam;
  double w0;
  double wa;
  double comoving_distance;
  int flag_stellarage;
  int flag_metals;
  unsigned int npartTotalHighWord[6];
  int flag_entropy_instead_u;
  int nothing;
  char fill[32];>---/* fills to 256 Bytes */
};
```

So `lenstools` inserts `w0`, `wa` and `comoving_distance` compared to the  `gadget2` header.

Any idea what happened? Regratably I found no version controlled version of `gadget2` to check if this was changed at some point.

**Comment by apetri** [comment]
We insterted these additional fields to allow a background expanding universe with a scale factor dynamics driven by dark energy. We were using a custom gadget version. I did this many years ago, but I believe LT allows for gaget I/O from the standard format as well

**Comment by uweschmitt** [comment]
Checked this. The only header implemented in `lenstools` is the modified version.

I'll close the merge request then.

**Comment by apetri** [comment]
Both headers are 256 bytes size, hence struct io_header_1 should be backward compatible. When you read a standard gadget header in a struct io_header_1 w0,wa are garbage but original gadget does not use them anyway... Anyway ok.

**Pull request closed without merging**